sycl/include/sycl/detail/group_sort_impl.hpp

```diff
@@ -436,23 +436,99 @@ template <> struct ValuesAssigner<false> {
   void operator()(IterOutT, size_t, ValT) {}
 };
 
+// Wrapper class for scratchpad memory used by the group-sorting
+// implementations. It simplifies accessing the supplied memory as arbitrary
+// types without breaking strict aliasing and avoiding alignment issues.
+struct ScratchMemory {
+public:
+  // "Reference" object for accessing part of the scratch memory as a type T.
+  template <typename T> struct ReferenceObj {
+  public:
+    ReferenceObj() : MPtr{nullptr} {};
+
+    operator T() const {
+      T value{0};
+      detail::memcpy(&value, MPtr, sizeof(T));
+      return value;
+    }
+
+    T operator++(int) noexcept {
+      T value{0};
+      detail::memcpy(&value, MPtr, sizeof(T));
```

Comment:
Can we use `sycl::bit_cast` here?

Reply:
I believe `sycl::bit_cast` requires the two types to be the same size. There is no guarantee that `sizeof(T) == sizeof(std::byte)` here.